tests/test_assets.py

```diff
@@ -1376,6 +1450,538 @@ def test_error_message_plurality(self,
+
+            # from 01 through 05 the second sid on the exchange held 'C'
+            with self.assertRaises(MultipleSymbolsFound):
+                finder.lookup_symbol('C', asof)
+
+            for n in range(num_exchanges):
+                C_result = finder.lookup_symbol(
+                    'C',
+                    asof,
+                    country_code=self.country_code(n),
+                )
+                assert_equal(
+                    C_result,
+                    finder.retrieve_asset(n * 2 + 1),
+                    msg=str(asof),
+                )
+                # The symbol and asset_name should always be the last held
+                # values
+                assert_equal(C_result.symbol, 'A')
+                assert_equal(C_result.asset_name, 'Asset A')
+
+        # no one held 'B' before 06
+        with self.assertRaises(SymbolNotFound):
+            finder.lookup_symbol('B', T('2014-01-05'))
+        for n in range(num_exchanges):
+            with self.assertRaises(SymbolNotFound):
+                finder.lookup_symbol(
+                    'B',
+                    T('2014-01-05'),
+                    country_code=self.country_code(n),
+                )
+
+        # no one held 'C' after 06, however, no one has claimed it yet
+        # so it still maps to sid 1
+        with self.assertRaises(MultipleSymbolsFound):
+            finder.lookup_symbol('C', T('2014-01-07'))
+
+        for n in range(num_exchanges):
+            assert_equal(
+                finder.lookup_symbol(
+                    'C',
+                    T('2014-01-07'),
+                    country_code=self.country_code(n),
+                ),
+                finder.retrieve_asset(n * 2 + 1),
+            )
+
+        for asof in pd.date_range('2014-01-06', '2014-01-11', tz='utc'):
+            # from 06 through 10 sid 0 held 'B'
+            # we test through the 11th because sid 1 is the last to hold 'B'
+            # so it should ffill
+            with self.assertRaises(MultipleSymbolsFound):
+                finder.lookup_symbol('B', asof)
+
+            for n in range(num_exchanges):
+                B_result = finder.lookup_symbol(
+                    'B',
+                    asof,
+                    country_code=self.country_code(n),
+                )
+                assert_equal(
+                    B_result,
+                    finder.retrieve_asset(n * 2),
+                    msg=str(asof),
+                )
+                assert_equal(B_result.symbol, 'B')
+                assert_equal(B_result.asset_name, 'Asset B')
+
+            # from 06 through 10 sid 1 held 'A'
+            # we test through the 11th because sid 1 is the last to hold 'A'
+            # so it should ffill
+            with self.assertRaises(MultipleSymbolsFound):
+                finder.lookup_symbol('A', asof)
+
+            for n in range(num_exchanges):
+                A_result = finder.lookup_symbol(
```

Comment:
It seems like there are a few patterns of assertions in this function that could be factored out to make them read more clearly, e.g.:

```python
def assert_resolves(symbol, asof, country_to_sid, expected_symbol, expected_name):
    with self.assertRaises(MultipleSymbolsFound):
        finder.lookup_symbol(symbol, asof)
    for n in range(num_exchanges):
        result = finder.lookup_symbol(symbol, asof, country_code=self.country_code(n))
        ...  # Other assertions
    
```